File: groupware/packages/net-utils/src/service-token.ts
/**
 * 서버 간 인증(Layer 3): X-Service-Token 발급/검증 (HS256)
 * 계약: docs/specs/service-http-contract.md §1.
 *
 * 표준 HS256 JWT 라 jsonwebtoken / Python hmac 구현과 상호호환된다(서명 입력, base64url 동일)
 * 런타임 의존성 0: node:crypto 만 사용
 */
import { createHmac, timingSafeEqual } from 'node:crypto';

const DEFAULT_TTL_SECONDS = 3600;

function b64url(data: string | Buffer): string {
  const buf = typeof data === 'string' ? Buffer.from(data) : data;
  return buf.toString('base64url');
}

export interface ServiceTokenPayload {
  service: string;
  iat: number;
  exp: number;
}

/** HS256 서비스 토큰 발급. jsonwebtoken.sign({service},secret,{HS256,1h}) 와 동일 산출 */
export function createServiceToken(
  secret: string,
  service: string,
  ttlSeconds: number = DEFAULT_TTL_SECONDS,
): string {
  const header = b64url(JSON.stringify({ alg: 'HS256', typ: 'JWT' }));
  const now = Math.floor(Date.now() / 1000);
  const payload = b64url(
    JSON.stringify({ service, iat: now, exp: now + ttlSeconds }),
  );
  const signingInput = `${header}.${payload}`;
  const signature = b64url(createHmac('sha256', secret).update(signingInput).digest());
  return `${signingInput}.${signature}`;
}

/**
 * 서비스 토큰 검증. 유효하면 payload, 아니면 null.
 * - 서명 불일치 / 형식오류 / exp 만료 → null
 * - allowed 가 주어지면 `service` 가 화이트리스트에 있어야 함
 */
export function verifyServiceToken(
  token: string,
  secret: string,
  allowed?: Iterable<string>,
): ServiceTokenPayload | null {
  const parts = token.split('.');
  if (parts.length !== 3) return null;
  const [headerB64, payloadB64, sigB64] = parts;

  const signingInput = `${headerB64}.${payloadB64}`;
  const expected = createHmac('sha256', secret).update(signingInput).digest();

  let actual: Buffer;
  try {
    actual = Buffer.from(sigB64, 'base64url');
  } catch {
    return null;
  }
  if (actual.length !== expected.length || !timingSafeEqual(actual, expected)) {
    return null;
  }

  let payload: ServiceTokenPayload;
  try {
    payload = JSON.parse(Buffer.from(payloadB64, 'base64url').toString('utf8'));
  } catch {
    return null;
  }

  if (typeof payload.exp !== 'number' || payload.exp < Math.floor(Date.now() / 1000)) {
    return null;
  }
  if (typeof payload.service !== 'string' || payload.service.length === 0) {
    return null;
  }
  if (allowed) {
    const set = allowed instanceof Set ? allowed : new Set(allowed);
    if (!set.has(payload.service)) return null;
  }
  return payload;
}

/**
 * SERVICE_TOKEN_SECRET 조회 (fail-closed)
 * prod 에서 미설정 시 throw, 비-prod 에서만 dev 폴백 허용
 */
export function resolveServiceSecret(
  value: string | undefined,
  opts: { isProduction: boolean; devFallback?: string } = { isProduction: true },
): string {
  if (value) return value;
  if (opts.isProduction) {
    throw new Error('SERVICE_TOKEN_SECRET 미설정: 서버 간 인증을 처리할 수 없습니다 (fail-closed).');
  }
  return opts.devFallback ?? 'dev-only-service-secret';
}

File: groupware/apps/api/fastapi/net-utils/src/csc_net_utils/service_token.py
"""서버 간 인증(Layer 3): X-Service-Token 발급/검증 (HS256).

표준 HS256 JWT 라 TS(`@csc/net-utils`) / jsonwebtoken 구현과 상호호환된다.
PyJWT 없이 표준 라이브러리(hmac/hashlib)만 사용한다.
계약: docs/specs/service-http-contract.md §1.
"""

from __future__ import annotations

import base64
import binascii
import hashlib
import hmac
import json
import time
from collections.abc import Iterable

_DEFAULT_TTL_SECONDS = 3600


def _b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def _b64url_decode(segment: str) -> bytes:
    padding = "=" * (-len(segment) % 4)
    return base64.urlsafe_b64decode(segment + padding)


def create_service_token(
    secret: str, service: str, ttl_seconds: int = _DEFAULT_TTL_SECONDS
) -> str:
    """HS256 서비스 토큰 발급. jsonwebtoken.sign({service},secret,{HS256}) 와 동일 산출."""
    now = int(time.time())
    header = {"alg": "HS256", "typ": "JWT"}
    payload = {"service": service, "iat": now, "exp": now + ttl_seconds}
    signing_input = (
        _b64url_encode(json.dumps(header, separators=(",", ":")).encode())
        + "."
        + _b64url_encode(json.dumps(payload, separators=(",", ":")).encode())
    )
    sig = hmac.new(secret.encode(), signing_input.encode(), hashlib.sha256).digest()
    return f"{signing_input}.{_b64url_encode(sig)}"


def verify_service_token(
    token: str, secret: str, allowed: Iterable[str] | None = None
) -> str | None:
    """토큰을 검증해 호출자(`service`)를 반환. 실패 시 None.

    - 서명 불일치 / 형식오류 / exp 만료 → None
    - allowed 가 주어지면 service 가 화이트리스트에 있어야 함
    """
    parts = token.split(".")
    if len(parts) != 3:
        return None
    header_b64, payload_b64, sig_b64 = parts

    signing_input = f"{header_b64}.{payload_b64}".encode()
    expected_sig = hmac.new(
        secret.encode() if isinstance(secret, str) else secret,
        signing_input,
        hashlib.sha256,
    ).digest()
    try:
        actual_sig = _b64url_decode(sig_b64)
    except (ValueError, binascii.Error):
        return None
    if not hmac.compare_digest(expected_sig, actual_sig):
        return None

    try:
        payload = json.loads(_b64url_decode(payload_b64))
    except (ValueError, binascii.Error):
        return None

    exp = payload.get("exp")
    if not isinstance(exp, (int, float)) or exp < time.time():
        return None

    service = payload.get("service")
    if not isinstance(service, str) or not service:
        return None
    if allowed is not None and service not in set(allowed):
        return None
    return service
